### PR TITLE
ECC cleanup

### PR DESCRIPTION
**Pull request by bbbrumley**
Here are three commits, somewhat related to each other. All follow from #6009.

802dae2  Removes ECDSA nonce padding. EC_POINT_mul is now responsible for constant time scalar multiplication, and hence the EC_GROUP that sets the wrapped function pointer. Wrt ECDSA, this prevents pretty much _every_ custom EC_METHOD from having to call BN_nnmod (leakage risk) to undo the higher level padding and put scalars back in the range [0,order).

1be6ae2 Deprecates ec2_mult.c. Instead of the group having a dedicated EC_POINT_mul, use the default one, unifying scalar multiplication code paths for ec2 and ecp generic curves. (Otherwise, you'd have to duplicate the nonce padding in ec2_mult.c)

831fa6d The EC_GROUP in ec2_smpl.c is not particularly efficient -- only supports affine coordinates. So there are lots of binary field inversions after the above commit. Due diligence: add some leakage countermeasure there. I also took the liberty of removing the OPENSSL_SUN_GF2M_DIV option -- sounds like an IPR minefield, no config switch enables it, it was written over 13 years ago when clmul didn't exist, etc. Alternative is to wrap *_div similarly, and it was much easier just to strip that #ifndef out ;)

Moving forward, wrt GF2 ECC, I see two options:

1. Eventually remove support entirely. If so, this is a step in that direction.
2. Enhance support. If so, you need projective coordinates for performance, which is easy to add. For example, rebase and tweak #2386 to be slightly more robust.

Tagging @romen @mattcaswell

### CODE REVIEW

**Comment by mattcaswell** [comment]
> Removes ECDSA nonce padding. EC_POINT_mul is now responsible for constant time scalar multiplication, and hence the EC_GROUP that sets the wrapped function pointer. Wrt ECDSA, this prevents pretty much every custom EC_METHOD from having to call BN_nnmod (leakage risk) to undo the higher level padding and put scalars back in the range [0,order).

So, #6009, made the `ec_wNAF_mul()` implementation call our new constant time routines. `EC_POINT_mul` calls `EC_POINTs_mul` which does this:

````
int EC_POINTs_mul(const EC_GROUP *group, EC_POINT *r, const BIGNUM *scalar,
                  size_t num, const EC_POINT *points[],
                  const BIGNUM *scalars[], BN_CTX *ctx)
{
    if (group->meth->mul == 0)
        /* use default */
        return ec_wNAF_mul(group, r, scalar, num, points, scalars, ctx);

    return group->meth->mul(group, r, scalar, num, points, scalars, ctx);
}
````
Are we confident that all non-default implementations (i.e. codepaths that don't end up calling `ec_wNAF_mul`)  do the right thing and won't be adversely impacted by this change?

**Comment by mattcaswell** [comment]
> I also took the liberty of removing the OPENSSL_SUN_GF2M_DIV option

I think this is ok - it doesn't impact our binary or source compatibility guarantee. Anything using that define (if anything exists) will still work. This probably means it is ok for backport to 1.1.0 too - although I'd like to hear if anyone else on @openssl/omc has an opinion.

WRT to backport to 1.1.0, this PR doesn't currently cherry-pick. I think the fix is trivial, but it would probably be best to create a separate PR specifically targeting 1.1.0.

**Comment by bbbrumley** [comment]
> Are we confident that all non-default implementations (i.e. codepaths that don't end up calling ec_wNAF_mul) do the right thing and won't be adversely impacted by this change?

I'm sure you get it but yea, that is why 1be6ae2019ffbaab727ccf1680826b2da6c6908b is related :) Here are the EC_METHODs:

```
$ grep '^const EC_METHOD' *.c
ec2_smpl.c:const EC_METHOD *EC_GF2m_simple_method(void)
ec_lib.c:const EC_METHOD *EC_GROUP_method_of(const EC_GROUP *group)
ec_lib.c:const EC_METHOD *EC_POINT_method_of(const EC_POINT *point)
ecp_mont.c:const EC_METHOD *EC_GFp_mont_method(void)
ecp_nist.c:const EC_METHOD *EC_GFp_nist_method(void)
ecp_nistp224.c:const EC_METHOD *EC_GFp_nistp224_method(void)
ecp_nistp256.c:const EC_METHOD *EC_GFp_nistp256_method(void)
ecp_nistp521.c:const EC_METHOD *EC_GFp_nistp521_method(void)
ecp_nistz256.c:const EC_METHOD *EC_GFp_nistz256_method(void)
ecp_smpl.c:const EC_METHOD *EC_GFp_simple_method(void)
```

@ekasper can speak to her various ecp_nistp*.c NIST implementations, but exec summary: yes.

Ditto for NISTZ from the Intel folks.

**Comment by richsalz** [comment]
I don't see any need to backport.  Congrats on removing the patent issue.

**Comment by mattcaswell** [comment]
> I don't see any need to backport. Congrats on removing the patent issue.

Are you commenting on the PR in general - or just the patent code removal bit? I think backporting the PR as a whole is highly desirable. If we feel we must keep the patent code then we might need to make some additional changes to `BN_GF2m_mod_div` in the backport version.

**Comment by richsalz** [comment]
I don't see a need to backport this PR.

**Comment by romen** [comment]
@mattcaswell just addedd some entries for the CHANGE file for the 3 commits here and PR #6009 

Are they ok?

I would also look for a place to document the shift that all these changes imply in the contract between implementers of EC_GROUP meths and callers of functions in EC_GROUP.

Can you think of a good spot where to add this documentation?

**Comment by bbbrumley** [comment]
> don't see a need to backport this PR.

@richsalz @mattcaswell I see both sides. Backporting #6009 made sense because it fixes a known vulnerability. Calling BN_nnmod to range reduce scalars -- no one has looked at that, publicly. Comments like

```
/*
 * this is an unusual input, and we don't guarantee
 * constant-timeness
 */
```

that I lifted from @ekasper are more of a "dunno what this callee does, so better at least document something" (IMHO).

File: CHANGES [hunk 1]
```diff
@@ -9,6 +9,26 @@
...
+     binary and prime elliptic curves.
+     [Billy Bob Brumley]
+
+  *) Remove ECDSA nonce padding: EC_POINT_mul is now responsible for
+     constant time fixed point multiplication.
+     [Billy Bob Brumley]
+
+  *) Revise elliptic curve scalar multiplication with timing attack
+     defenses: ec_wNAF_mul redirects to a constant time implementation
+     when computing fixed point ad variable point multiplication (which
```
Comment (mattcaswell) [inline CHANGES]: typo: and

**Comment by mattcaswell** [comment]
@romen:

> I would also look for a place to document the shift that all these changes imply in the contract between implementers of EC_GROUP meths and callers of functions in EC_GROUP.
>
> Can you think of a good spot where to add this documentation?

Callers of functions should be using the public API which is all documented, e.g. see:

https://github.com/openssl/openssl/blob/master/doc/man3/EC_POINT_add.pod

So if there are any public API changes then that's where we should document them. But I'm thinking from a caller perspective not much has changed, except to say the padding of the scalar input for `EC_POINT_mul` is no longer required - but we never said it was in the first place (and if anyone is doing it now and continues to do it afterwards then, that's unnecessary but - no big deal). Any public API change we make must be 100% backwards compatible.

EC_METHOD is actually a purely internal concept (we don't enable any mechanism for users to create their own EC_METHOD implementation). It's a private structure with no public constructor functions. So any documentation for implementers is internal documentation only - which is to say we add it as code comments in appropriate places. Perhaps in the structure definition itself?

**Comment by mattcaswell** [comment]
@bbbrumley what is the status of this? I note there is an outstanding minor nit, and a suggestion from @romen that some documentation is coming.

**Comment by romen** [comment]
@mattcaswell sorry, I lost track of the comment on the typo!

Will fix both today!

File: crypto/ec/ec_lcl.h [hunk 2]
```diff
@@ -120,6 +120,23 @@ struct ec_method_st {
...
+     *        + points[num-1] * scalars[num-1].
+     *
+     * For a fixed point multiplication (scalar != NULL, num == 0)
+     * or a variable point multiplication (scalar == NULL, num == 1),
+     * mul() must use a constant time algorithm: in both cases callers
+     * should provide an input scalar (either scalar or scalars[0])
+     * in the range [0, ec_group_order); for robustness, implementers
+     * should handle the case when the scalar has not been reduced, but
+     * may treat it as an unusual input, without any constant-timeness
+     * guarantee.
```
Comment (romen) [inline crypto/ec/ec_lcl.h]: @mattcaswell 

> But I'm thinking from a caller perspective not much has changed, except to say the padding of the scalar input for `EC_POINT_mul` is no longer required - but we never said it was in the first place (and if anyone is doing it now and continues to do it afterwards then, that's unnecessary but - no big deal). Any public API change we make must be 100% backwards compatible.

If callers keep doing the padding it is not only unnecessary, but constant-timeness is not guaranteed, which is a big deal once we delegate constant-timeness  for fixed and variable point multiplication to `EC_POINT_mul`.

This is actually the change that we wanted to make explicit!

It is 100% backwards binary compatible with existing alternative `EC_METHODs` (cf. https://github.com/openssl/openssl/pull/6070#issuecomment-384311256 and https://github.com/openssl/openssl/pull/6070#issuecomment-384248024), but needs to be stated for future `EC_METHOD` implementations.

**Comment by dot-asm** [comment]
> I don't see a need to backport this PR.

I for one would argue in favour of removal of specifically nonce padding even in 1.0.2...

**Comment by romen** [comment]
I'm in the process of squashing together the documentation commits with the relevant code commits, I plan to reduce the 10 current commits to 4 (3 about the 3 different changes of this PR, a 4th about the documentation covering this and the previous PR)

**Comment by romen** [comment]
rebase done.

@dot-asm @mattcaswell @levitte  Do you need to re-approve it?

**Comment by dot-asm** [comment]
Merged. Thanks! Closing. Though there was no "tie-break" on whether or not it should be backported, there only two opposite suggestions...

**Pull request closed without merging**